### ISSUE TITLE
Page deletion has no effect when saving the PDF

### ISSUE THREAD

**Issue by kobaltcore** [pypdfium2-team/pypdfium2#96]
I've been trying to move to pdfium from mupdf -largely due to the licensing issues with the latter- and in the process of porting over some of my code I've run into a strange issue in regards to deleting pages from PDF documents.
As far as I can tell the utility classes and methods do not offer a helping hand with this, so I used the low-level API instead.
It could very well be that I'm using this wrong, but to me it appears that while *something* is being deleted, it has no bearing on the output document.
Is there a specific magic incantation I have to produce to get this to work in addition to the example below?

# Minimal Example
```py
import pypdfium2 as pdfium

# load in the test document (any PDF with multiple pages will do)
doc = pdfium.PdfDocument("in.pdf")
# page count is correct
print("   before page deletion", pdfium.FPDF_GetPageCount(doc.raw))

# delete the first page
pdfium.FPDFPage_Delete(doc.raw, 0)
# page count is correct, one less than before
print("    after page deletion", pdfium.FPDF_GetPageCount(doc.raw))

# save the document
with open("out.pdf", "wb") as f:
    pdfium.save_pdf(doc.raw, f)

# load the saved document
doc = pdfium.PdfDocument("out.pdf")

# page count is incorrect, same as the original document
print("loading edited document", pdfium.FPDF_GetPageCount(doc.raw))
```

In general I would be interested in a couple more examples of working with this API, especially in the realm of moving about and modifying pages, i.e. deleting them, adding new pages from other PDF's (which is somewhat covered by the PDF merge example) as well as creating new pages from an image file.

**Comment by mara004** [comment]
Thanks for the clear report. As far as I can see, your sample code is correct and I am able to reproduce the issue. However, I currently don't have an idea what could possibly be going on in this case. I'll have to investiagte this in more detail.

Of course you're right there should be more examples - I'll hope to work on this at some point, and I'm also happy to take new examples from contributors. That said, I can recommend to look at the `pdfbrain` project which contains some good examples on how to work with the PDFium API.
My main use for pypdfium2 currently is rendering and outline parsing, so I did not have much time to look into other parts of the API yet. I personally prefer to use pikepdf for structural PDF transformations like moving pages.

**Comment by mara004** [comment]
In the meantime, I've come up with a slighly different approach how you can still achieve the goal of deleting pages. It can be done by initialising a new document and copying over the pages to keep rather than deleting pages of an existing document:

```python3
import pypdfium2 as pdfium
import ctypes

# load pdf
src_pdf = pdfium.PdfDocument("in.pdf")

# create new document
dest_pdf = pdfium.FPDF_CreateNewDocument()

# copy pages into a new document rather than deleting pages of an existing one
page_count = pdfium.FPDF_GetPageCount(src_pdf.raw)
pdfium.FPDF_ImportPages(dest_pdf, src_pdf.raw, "1-3", 0)
assert pdfium.FPDF_GetPageCount(dest_pdf) == 3

# save the new document
with open("out.pdf", "wb") as fh:
    pdfium.save_pdf(dest_pdf, fh)

src_pdf.close()
pdfium.close_pdf(dest_pdf)

# reopen it to check the page count
reopened_pdf = pdfium.PdfDocument("out.pdf")
assert pdfium.FPDF_GetPageCount(reopened_pdf.raw) == 3
reopened_pdf.close()
```

**Comment by mara004** [comment]
I think this is most likely a PDFium issue. A cause from python/ctypes is very improbable from what I can tell.
However, upstream does not take bug reports seriously unless they come with a minimal reproducible example written in C/C++.
@kobaltcore Are you in a position to create such an example? I'm sorry to say I can't write C code personally.

**Comment by kobaltcore** [comment]
Update: I figured out what the issue is, and it's not pdfium itself.

Instead it boils down to an oversight in the saver you implemented here:
https://github.com/pypdfium2-team/pypdfium2/blob/3acc962190b57c7c188569a65258cb5a6380c143/src/pypdfium2/_helpers/saver.py#L44

The crux of the issue is the `FPDF_INCREMENTAL` flag. This causes pdfium to save changes to the document, which is what it should do, but since it's incremental, it only saves additions and modifications, *not* deletions. If you switch to `FPDF_NO_INCREMENTAL` it works as expected because it's now able to use destructive operations.

I did make a sample in C anyway (before I figured this out) so I'll include it here for posterity.
```c
#include <stdio.h>
#include <fpdf_edit.h>
#include <fpdf_save.h>

FILE* pdfFile;

int WriteBlockImpl(FPDF_FILEWRITE* pThis, const void* pData, unsigned long size)
{
    fwrite(pData, size, 1, pdfFile);
    return 1;
}

int main() {
  FPDF_LIBRARY_CONFIG config;
  config.version = 2;
  config.m_pUserFontPaths = NULL;
  config.m_pIsolate = NULL;
  config.m_v8EmbedderSlot = 0;

  FPDF_InitLibraryWithConfig(&config);

  // create handle to output doc
  pdfFile = fopen("out.pdf", "wb");

  // load in the test document (any PDF with multiple pages will do)
  FPDF_STRING test_doc = "test.pdf";
  FPDF_DOCUMENT doc = FPDF_LoadDocument(test_doc, NULL);

  // page count is correct
  printf("before page deletion %d\n", FPDF_GetPageCount(doc));

  // delete the first page
  FPDFPage_Delete(doc, 0);
  // page count is correct, one less than before
  printf(" after page deletion %d\n", FPDF_GetPageCount(doc));

  // save document
  FPDF_FILEWRITE fileWriteObj;
  fileWriteObj.version = 1;
  fileWriteObj.WriteBlock = WriteBlockImpl;

  FPDF_SaveAsCopy(doc, &fileWriteObj, FPDF_NO_INCREMENTAL);

  FPDF_CloseDocument(doc);

  FPDF_DestroyLibrary();

  return 0;
}
```

**Comment by mara004** [comment]
Oh yes, absolutely! Thanks a lot for finding this out. I'll push a commit to fix this as soon as possible.

**Comment by mara004** [comment]
It looks like the saving flags are not documented - that might be something to address upstream.

**Comment by kobaltcore** [comment]
I did find them [here](https://developers.foxit.com/resources/pdf-sdk/c_api_reference_pdfium/group___f_p_d_f_i_u_m.html#ga1d5b1a39fb70282aecc2f1a45a68bf5c), though the information contained therein is not particularly expansive.

**Comment by mara004** [comment]
They're listed, yes - but if the description is essentially the same as the variable name, that does not really help.

**Comment by kobaltcore** [comment]
True. Alas, I don't recall anymore where exactly I found the description of these flags. I did happen across them during my adventures in porting this example to C, but unfortunately I forgot. Was probably buried somewhere deep in the source code. Either way, it's definitely a bit of a shame that the pdfium docs are as sparse as they are.

**Comment by mara004** [comment]
Before I make the release tomorrow, I would like to make sure the behaviour of `save_pdf()` is correct.
Do you know whether it is sufficient to use `FPDF_NO_INCREMENTAL`, or should I combine both flags, like so: `FPDF_NO_INCREMENTAL | FPDF_INCREMENTAL`?

**Comment by kobaltcore** [comment]
As far as I understand these are mutually exclusive, so they should not be OR'd together. OR'ing them would result in `1 | 2` which equals `3` and thus maps to `FPDF_REMOVE_SECURITY`, which is not the intended save flag. As such, just leaving it at `FPDF_NO_INCREMENTAL` should be sufficient.

**Comment by mara004** [comment]
All right, thanks for the clarification!

**Comment by mara004** [comment]
This issue should be resolved since version 1.3.0, which I released today. Thanks again for finding and fixing it!
